From 68b846fb1f296cad747a266eef4778ced3fa30d6 Mon Sep 17 00:00:00 2001
From: Matthew Garrett <mjg59@google.com>
Date: Sat, 11 Apr 2020 16:17:24 -0700
Subject: [PATCH 026/134] Add an adaptive target callback into RAPL code

Add support for setting RAPL parameters based on adaptive targets.
Incomplete at present.

(cherry picked from commit ea666b8865fe93b39dea75b53ba4de5f4e9ab76d)
Signed-off-by: Paolo Pisati <paolo.pisati@canonical.com>
(cherry picked from commit 0603324d7b8a4992ebc402debe4c31ab4eee59ec)
Signed-off-by: Paolo Pisati <paolo.pisati@canonical.com>
---
 src/thd_cdev_rapl.cpp | 9 +++++++++
 src/thd_cdev_rapl.h   | 1 +
 2 files changed, 10 insertions(+)

diff --git a/src/thd_cdev_rapl.cpp b/src/thd_cdev_rapl.cpp
index 51b6c95..dcee3da 100644
--- a/src/thd_cdev_rapl.cpp
+++ b/src/thd_cdev_rapl.cpp
@@ -265,6 +265,15 @@ int cthd_sysfs_cdev_rapl::rapl_read_enable_status()
 
 }
 
+void cthd_sysfs_cdev_rapl::set_adaptive_target(struct adaptive_target target) {
+	int argument = std::stoi(target.argument, NULL);
+	if (target.code == "PL1MAX") {
+		set_curr_state(argument * 1000, 1);
+	} else if (target.code == "PL1TimeWindow") {
+		rapl_update_time_window(argument * 1000);
+	}
+}
+
 int cthd_sysfs_cdev_rapl::update() {
 	std::stringstream temp_str;
 	int constraint_phy_max;
diff --git a/src/thd_cdev_rapl.h b/src/thd_cdev_rapl.h
index bafbf1f..f2070bd 100644
--- a/src/thd_cdev_rapl.h
+++ b/src/thd_cdev_rapl.h
@@ -96,6 +96,7 @@ public:
 	virtual int get_max_state();
 	virtual int update();
 	virtual void set_curr_state_raw(int state, int arg);
+	void set_adaptive_target(struct adaptive_target target);
 	void thd_cdev_set_min_state_param(int arg);
 	int get_phy_max_state() {
 		return phy_max;
-- 
2.27.0

